### PR TITLE
feat: add tool for open-set detection using grounding dino

### PR DESCRIPTION
**Pull request by rachwalk**
## Purpose

Adds Grounding Dino Tooling for distance estimation. Grounding Dino is used to extract the regions of the camera POV classified as certain objects. These regions can then be extracted from the depth channel, and after calibration, used to estimate distance from camera to them.

## Proposed Changes

- Adds two tools: `GetDetectionTool`, `GetDistanceToObjectsTool`.
- Defines API for the `grounding_dino` extension

## Issues

https://github.com/RobotecAI/rai/issues/194

## Testing

Tested using following script:

```
import rclpy
from rai_grounding_dino.tools import GetDetectionTool

from rai.node import RaiBaseNode

if __name__ == "__main__":
    rclpy.init()
    node=RaiBaseNode(node_name="unique_node")
    x = GetDetectionTool(node)._run(
        camera_topic="/camera/camera/color/image_raw",
        object_names=["chair"],
    )
    print(f"Detected: {x}")
    x = GetDistanceToObjectsTool(node)._run(
        camera_topic="/camera/camera/color/image_raw",
        depth_topic="/camera/camera/depth/image_raw_rect",
        object_names=["chair"],
    )
    print(f"Detected: {x}")
```

In second terminal run:

```
ros2 launch rai_grounding_dino gdino_launch.xml
```

Camera topics must be available and of the same resolution. The detection went through correctly.

### CODE REVIEW

**Comment by adamdbrw** [comment]
Could you include the test in the commit, and document the new tool as well for users and developers? Does it require additional dependencies?

File: src/rai_interfaces/msg/RAIDetectionArray.msg [hunk 1]
```diff
@@ -20,4 +20,5 @@ std_msgs/Header header
 # A list of the detected proposals. A multi-proposal detector might generate
 #   this list with many candidate detections generated from a single input.
 vision_msgs/Detection2D[] detections
+# a list of classes detected
```
Comment (rachwalk) [inline src/rai_interfaces/msg/RAIDetectionArray.msg]:
```suggestion
# a list of classes being detected
```

File: src/rai_extensions/rai_grounding_dino/rai_grounding_dino/tools.py [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import List, NamedTuple, Optional, Type
+
+import numpy as np
+import rclpy
+import sensor_msgs.msg
+from langchain_core.pydantic_v1 import Field
```
Comment (maciejmajek) [inline src/rai_extensions/rai_grounding_dino/rai_grounding_dino/tools.py]:
```suggestion
from pydantic import Field
```

File: src/rai_extensions/rai_grounding_dino/rai_grounding_dino/tools.py [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+    ):
+        camera_img_msg = self._get_image_message(camera_topic)
+        future = self._call_gdino_node(camera_img_msg, object_names)
+
+        while rclpy.ok():
+            resolved = self._spin(future)
+            if resolved is not None:
+                detected = self._parse_detection_array(resolved)
+                names = ", ".join([det.class_name for det in detected])
+                return f"I have detected the following items in the picture {names}"
```
Comment (maciejmajek) [inline src/rai_extensions/rai_grounding_dino/rai_grounding_dino/tools.py]:
```suggestion
                return f"I have detected the following items in the picture {names or 'None'}"
```

File: src/rai_extensions/rai_grounding_dino/rai_grounding_dino/tools.py [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+                measurements = self._get_distance_from_detections(
+                    depth_img_msg, detected, threshold, conversion_ratio
+                )
+                measurement_string = ", ".join(
+                    [
+                        f"{measurement[0]}: {measurement[1]:.2f}m away"
+                        for measurement in measurements
+                    ]
+                )
+                return f"I have detected the following items in the picture {measurement_string}"
```
Comment (maciejmajek) [inline src/rai_extensions/rai_grounding_dino/rai_grounding_dino/tools.py]:
```suggestion
                return f"I have detected the following items in the picture {measurement_string or 'no objects'}"
```

**Review by maciejmajek — COMMENTED** [review]
Thanks. I can confirm that the feature works as expected. Please take a look at the comments. 
Regarding the {... or 'Nothing'}, decide yourself whether handling it that way or using if measurements[] is better

**Review by maciejmajek — APPROVED** [review]

**Pull request merged**